### ISSUE TITLE
Fix `declaration-property-value-no-unknown` false negatives for `position-try-options`

### ISSUE THREAD

**Issue by Mouvedia** [stylelint/stylelint#7964]
### What minimal example or steps are needed to reproduce the bug?

```css
a {
  position-try: foo;
  position-try-fallbacks: foo;
  position-try-options: foo;
  position-try-order: foo;
}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "rules": { "declaration-property-value-no-unknown": true }
}
```

### How did you run Stylelint?

[demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilAHCAzgJYAuREmAtCQE4Ce8KAZhBANxY77FkXX2UmSADbCARkjABrAoxbtOeQqXJVadShFy9Ms5qw7YlPVfw0QasGnINYAviAA0IJkWEwAckgC2cRDAAPH1x3ADowAgIncApXAHMEEED8GhIUWCEAV2E0jCN0EBpsmAICxmAUAtgwYSQaJB1KXBotGFSNADcRTJhKTAhKTMwpfoB3TDKUWh6UO3ssaMhMeIAxC28GxIISOndhIkwScNiiONDvACso51hcKMQ8nALt3Zh9w8mC2pISkgLHRTPHZ7A4kShLeKUbZITBQOpQT4gb6-Ar2EB2IA)

### What did you expect to happen?

`Unexpected unknown value "foo" for property "position-try-options"`

### What actually happened?

`✅ No problems!`

### Resources

mdn/browser-compat-data#23651 w3c/csswg-drafts#10395
https://caniuse.com/?search=position-try

**Comment by Mouvedia** [comment]
In light of [this announcement](https://github.com/stylelint/stylelint/pull/7957#issuecomment-2309655378), I am not gonna set the status to _ready to implement_ right away.

**Comment by romainmenke** [comment]
To illustrate the scale of the issue.

There have been between 600 and 700 different additions and modifications to CSS syntax definitions since `csstree` was last updated and published.

This is not something we can manually rectify here in Stylelint :)
This is going to take a while

**Comment by Mouvedia** [comment]
> This is not something we can manually rectify here in Stylelint :)

We could but we shouldn't.
It would just be a matter of maintaining one or several JSON files.
Clearly this wouldn't be ideal but it would be better than having to fork `css-tree`.

**Comment by romainmenke** [comment]
I can only speak for myself, but I can definitely say that maintaining a list of hundreds of highly specific patches for complex definitions and doing this by hand is beyond what I can and am willing to do.

This work will never end, and it will likely increase over time.

It would not just be a matter of maintaining some JSON files.
Maintaining those JSON files is the issue, not the solution.

If this were paid work, then maybe I would consider it, but I am not going to spend hours every month, for free, on this type of work. It is the kind of work that only takes and gives nothing back. There is nothing that can be learned from it, no challenge in it, ...

Acknowledging the situation and the massive scope of the effort would make me feel more supported than casually stating that this is something you could just do 😛

-------

Anyway, if I am not successful in finding ways to automate this process then I think we should discuss removing the rule.

**Comment by Mouvedia** [comment]
> Anyway, if I am not successful in finding ways to automate this process then I think we should discuss removing the rule.

I think that as long it's not part of standard or worse recommended configs it could be kept.
The drawbacks were known when we added it, and I wish you could have participated in our discussion in #6511.
Maybe you would have been on my side, warning about problems that back then seemed improbable.
Now that it's been added, I consider that it's our responsibility to help the rule graduate from its experimental stage.
We owe it to the users. I see that you don't have the same stance on the matter and that's fine: we don't have to agree on everything. And your opinion is actually more pragmatic which will probably result in a steadier pace for us maintainers in the long run.

> This rule is experimental with some false negatives that we'll patch in minor releases.

**Comment by ybiquitous** [comment]
I'm against removing the `declaration-property-value-no-unknown` rule, but it's worth discussing whether it should be removed from our recommended config. In my opinion, this rule is still valuable for many people because I guess they don't often use brand-new properties.

If patch auto-generation for `css-tree` doesn't work, we may need to add ad-hoc changes manually for a while when people open a bug report.

**Comment by Mouvedia** [comment]
> but it's worth discussing whether it should be removed from our recommended config

No need: it's not part of any of our configs.

> when _people_ open a bug report.

I guess a maintainer shouldn't be proactive about it then.
Let's say that from on, only tickets opened by users or false positives can be fixed.
i.e. I cannot open false negatives excluding this very issue
This should reduce our burden.

**Comment by ybiquitous** [comment]
> No need: it's not part of any of our configs.

Ah, sorry. I misunderstood.

**Comment by Mouvedia** [comment]
It seems like there's a precedent regarding forks of csstree repositories.
[`csstree/stylelint-validator` was forked](https://github.com/carlosjeurissen/stylelint-validator/tree/stylelint-16) back in February.

**Comment by romainmenke** [comment]
I am also against removing this rule, but I am trying to balance various nuances and needs :)

I don't think it should be controversial to say that there are limits to how much workload we should volunteer to take on.

We don't have any obligation to users of Stylelint. That a rule was once created does not mean that we need to keep maintaining it. Open source is a gift, we don't owe anything to anyone.


All the stylistic rules were also offloaded.

-----

A rule that is effectively frozen and warns on invalid syntax only because it is outdated is also not helpful to users.

If the data source is patched incorrectly we will cause more false errors than anything else.

The challenge with this rule is that the CSS syntax is a very large, hand written, cyclic graph.

Any change can affect many other types, including itself when it is cyclic definition.

Mutating this by hand, without proper tests, is extremely error prone:
- you might have broken downstream definitions
- you might subtly affect groupings

Mutating this by hand is extremely hard to maintain.

**Comment by romainmenke** [comment]
I am also against removing this rule, but I am trying to balance various nuances and needs :) If I were part of the original discussion I would have been fully in favor of adding it and I still am today.

I think this rule is incredibly valuable.

--------

I don't think it should be controversial to say that there are limits to how much workload we should volunteer to take on.

Open source is a gift, we don't owe anything to anyone. That a rule was once created does not mean that we need to keep maintaining it. This doesn't mean that I don't care about Stylelint users or the quality and future of Stylelint. It only means that we don't **owe** anyone our time, focus, ...

Even if some are willing to do more work now, we need a long term commitment for efforts like these. I am concerned that the commitment required to do rule right is beyond what is healthy for Stylelint.

--------

I am not saying that I want this rule removed.

I am only saying that we should discuss this if my effort prove unsuccessful.
I also want us to be careful with the amount of work we volunteer to do.

-----

The challenge with this rule is that the CSS syntax is a very large, hand written, cyclic graph.

Any change can affect many other types, including itself when it is cyclic definition.

Mutating this by hand, without proper tests, is error prone:
- you might break downstream definitions
- you might subtly affect groupings
- you might introduce a naming conflict

Mutating this by hand is hard to maintain:
- you don't get any signals about which overrides are outdated themselves
- you don't get any signals when upstream data includes your overrides or even newer syntax
- you don't get any signals when overrides break the cyclic graph structure
- you don't get any signals on naming conflicts
- you don't get any signals on orphaned types
- you don't get any signals on missing types

The CSSWG does not really consider these syntax definitions as part of their "API surface".
So they often rewrite these drastically. Removing and introducing dozens of new definitions.

Any change is also potentially a breaking change for users.
If a type is altered or removed but users are using these same types in their own overrides, then we are breaking their config.

-----

I am making slow progress here: https://github.com/csstools/postcss-plugins/pull/1466

Things I haven't started yet:
- `atrule` descriptors and their syntax
- orphaned definition analysis
- test coverage, fuzzzing, mutators

Does seem to work ok:
- record dependencies between various specs and order them correctly
- combining and overriding conflicting definitions across delta specs
- generating and re-generating patches that need to be merged by hand
- testing that syntax definitions are valid
- testing sample bits of CSS against syntax definitions
- parsing and validating various community CSS files (bootstrap, ...)

With each update to `@webref/css` the build script will remove outdated patch entries and insert new template patch entries. These patch entries need to be manually resolved.

A patch entry looks like this:

```json
	"ray()": {
		"type": "added",
		"syntax-before": "",
		"syntax-after": "ray( <angle> && <ray-size>? && contain? && [ at <position> ]? )",
		"comment": "Without the extra group the order seems to matter anyway, possibly a bug in css-tree or the syntax definition",
		"merged": "ray( [ <angle> && <ray-size>? && contain? && [ at <position> ]? ] )",
		"tests": {
			"passing": [
				{
					"property": "offset-path",
					"value": "ray(45deg)"
				},
				{
					"property": "offset-path",
					"value": "ray(200deg contain)"
				},
				{
					"property": "offset-path",
					"value": "ray(contain 200deg)"
				},
				{
					"property": "offset-path",
					"value": "ray(50deg closest-corner contain at 100px 20px)"
				},
				{
					"property": "offset-path",
					"value": "ray(50deg closest-corner contain)"
				}
			],
			"failing": [
				{
					"property": "offset-path",
					"value": "ray()"
				},
				{
					"property": "offset-path",
					"value": "ray(50deg closest-corner contain at)"
				},
				{
					"property": "offset-path",
					"value": "ray(50hz)"
				}
			]
		}
	},
```

The entire setup focusses heavily on creating a process that makes it possible to understand what changes, which definitions are affected and if resolved patches are actually valid.

If anyone is willing to contribute I will gladly explain more and setup an example.
There are more than 700 patches still to resolve so plenty todo 😅 

I am currently estimating that this will take about 2-3 months to complete.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the elaboration. I now understand how this work is challenging.

> With each update to `@webref/css` the build script will remove outdated patch entries and insert new template patch entries. These patch entries need to be manually resolved.

I was very surprised to see the comment above... 😅

**Comment by romainmenke** [comment]
`position-try-options` was renamed to `position-try-fallbacks`: https://github.com/w3c/csswg-drafts/pull/10517

And `position-try-fallbacks` is supported in `csstree`.

I think this issue is resolved.

**Comment by Mouvedia** [comment]
I disagree. We ought to support all properties that were implemented at one point by a browser.
i.e. Chrome/Edge 125–127 and Opera 111

Since it's not yet a policy it can be discussed but it shouldn't be closed.

**Comment by romainmenke** [comment]
I think it would be better if Stylelint doesn't have its own opinion on what is valid syntax.

What is or isn't syntactically valid should be defined by the CSSWG, not by Stylelint.

The data source for what is valid should be something actively maintained by the W3C and CSSWG, not Stylelint. (i.e. `mdn-data` and `@webref/css`)

`mdn-data` and `csstree` are already really conservative and both have editorial escape hatches to preserve non-standard properties or values in their data sets.

`mdn-data` and `csstree` decided not to preserve `position-try-options`.
I don't think Stylelint should swim against the current here.

-----

Having a clear policy that any data issues are upstream issues would be best in my opinion.
Anyone who disagrees with the data can still file an upstream issue :)

-----

I think there is a very large difference between wanting to compensate for `csstree` moving to slowly and marking currently invalid CSS as valid because it was present in 3 browser versions of a single vendor as an experimental feature.

Users want to know when they write something that is invalid and `position-try-options` today is invalid CSS.

**Comment by ybiquitous** [comment]
@romainmenke's thought makes sense to me.

I don't think Stylelint should actively support removed or deprecated CSS features unless they don't widely impact many people.

**Comment by romainmenke** [comment]
Small status update on https://github.com/csstools/postcss-plugins/pull/1466

Properties and types are done and can be seen here: https://github.com/csstools/postcss-plugins/blob/a02ec16572375dbf1c211d157d765a76c74d1359/packages/css-syntax-patches-for-csstree/dist/index.json

Starting work on atrule descriptors next.

**Comment by romainmenke** [comment]
I am unsure what the status of this issue is.

This seems to work as expected: [Stylelint Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilAHCAzgJYAuREmAtCQE4Ce8KmENAtkgDYoBmHRulAEYcIYANYBuLDnzEyFavUrdOHQUnEFGvfkJHip2PIVLkqtOpRawajVoRKUA7kSgkAFoYC+WLILTSxnJmigw8EBCGMiby5koqHGoaYlrhkYGypgoWVjQ22hHeIAA0INxEHDAAckiscIgwAB61uJUAdGAEBCXgFOUA5gggTfg0JCiwKgCuHOMYRuggNDMwBIuMwCiLsGAcSDRIsZS4NBC4MGOWAG6cUzCUzJRTmGLMTpjrKLR3KD6Yfz1IJgBgAxFjsEhDAgkOiVPiYEgdPpEfptVgAK26pVguG6iHmOEW0NhMHhJE+iz2JFW5JKgSJMLhRARlCBA0o0KQmCg+ygFJAVJpiywXhAXiAA)

`position-try-options` is maybe a good candidate for `property-no-deprecated` ?

**Comment by Mouvedia** [comment]
> `position-try-options` is maybe a good candidate for `property-no-deprecated`?

@romainmenke 

https://github.com/stylelint/stylelint/blob/be7a5b6b1f540759f691f2ec47422699a6343ae4/lib/rules/property-no-deprecated/index.mjs#L60

**Comment by romainmenke** [comment]
@Mouvedia then this issue can be closed?

**Comment by Mouvedia** [comment]
> then this issue can be closed?

I don't see `overflow-x`/`overflow-x`/`position-try-options` in https://github.com/csstools/postcss-plugins/blob/main/packages/css-syntax-patches-for-csstree/dist/index.json so no not yet.

i.e. https://github.com/csstree/csstree/commit/9558ba790daeda2b24935838bf89990699ece66e was merged _after_ the last release of `css-tree`

> [In the future more groupings might be added.](https://github.com/csstools/postcss-plugins/tree/main/packages/css-syntax-patches-for-csstree)

```
	'overflow-x': '| <-non-standard-overflow>',
	'overflow-y': '| <-non-standard-overflow>',
```
belong in `next` since it already has `"overflow": "<'overflow-block'>{1,2} | <-non-standard-overflow>",` .
i.e. it's an oversight

`position-try-options` is an alternate name of `position-try-fallbacks` but I don't know your policy regarding aliases.
i.e. are the new names replacing the older ones?
The README doesn't say if you accept manual additions.
i.e. if the files are later auto-regenerated will the manual additions be removed?

**Comment by romainmenke** [comment]
Hi @Mouvedia honestly I almost regret asking about this issue 😕 

I spend the last 15 minutes writing a detailed reply to each part of your comment but I ended up deleting all of it. I felt like I was just repeating things that I have already said before and I don't want to keep going in circles.

I welcome contributions in `@csstools/css-syntax-patches-for-csstree` but the goal of that package is to have a solid definition of valid CSS syntax. Something that was only valid for a few versions in a single engine does not meet that bar.

-----

`overflow-y`/`overflow-x` have their own issue and this is off topic here.

**Comment by Mouvedia** [comment]
> Something that was only valid for a few versions in a single engine does not meet that bar.

That's understandable.

> I felt like I was just repeating things that I have already said before […]

Right, I have read your previous comments and I see now how my questions were already answered.
e.g. "These patch entries need to be manually resolved."

> `overflow-y`/`overflow-x` have their own issue and this is off topic here.

Indeed. [This commit](https://github.com/csstree/csstree/commit/7289677d72b32b69dc49491c86ac547ad8e2f80a) will be part of the next release.

> I welcome contributions in `@csstools/css-syntax-patches-for-csstree` […]

Great.

**Comment by romainmenke** [comment]
Fixed upstream in `css-tree`: [Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdilAHCAzgJYAuREmAtCQE4CelEuZFB8KAZhBANxYC+IADQgORADYwAckgC2cRDAAec3JIB0YAgWHgKYgOYIQy-DRIpYHJAFdxFjNhToQNOzAIv2wZyFhhxJBokFipcGiYYcwYANyRxGxhKTAhKG0wAaxSAd0wvFFpElH4BLF1ITEMAMQgaWRDjAhI6SXEiTBJNfSIDdVkAKx0RWFwdREccFyaWmDaO-JdAkg8SFyEsSZBp1vaSSgrDSiakTCggqAWQJZWXARB+IA)
